documentation/sphinx/source/rangelock.rst

```diff
@@ -0,0 +1,122 @@
+###################################
+FoundationDB Per-Range Lock (WiP)
+###################################
+
+| Author: Zhe Wang
+| Reviewer: Jingyu Zhou
+| Audience: FDB developers, SREs and expert users.
+
+
+Overview
+========
+Per-Range Lock is a feature that allows to shut down write traffic to a specific range of keys in FoundationDB (FDB).
+The locked range must be within the user key space, aka ``"" ~ \xff``.
+If a user/app grabs a lock on a range, other user/app can read the range but cannot write to the range. Essentially, this lock type is
+a readLock. The read lock is not exclusive --- A range can have multiple locks by different users. 
+
+On the other hand, there is a concept of write lock in the context of FDB per-range lock --- If a user/app takes a writeLock on a range, 
+other user/app cannot do any read nor write. The write lock is exclusive, if a user/app takes the writeLock on a range, the range must have not any other lock.
+Currently, we only implemented the readLock. The writeLock is currently not implemented. we will implement the writeLock later on demand. 
+
+Example use cases
+-----------------
+Currently, BulkLoad feature is an example of using the per-range lock. 
+BulkLoad is developed to load a large amount of data into a FDB range without going through the transaction system.
+As a result, BulkLoad must ensure the correctness during the data injection in the presence of user traffic. 
+To achieve this, BulkLoad locks the range that it is going to write to, and then it writes the data to the range.
```

Comment:
```suggestion
To achieve this, BulkLoad locks the range to prevent user traffic from writing to the database, and then loads the range data directly into storage servers.
```